graphql_compiler/compiler/workarounds/orientdb_query_execution.py

```diff
@@ -273,6 +273,8 @@ def _expose_only_preferred_locations(match_query, location_types, coerced_locati
                         # or has already been applied at a previous occurrence of this location.
                         # We can simply delete the QueryRoot / CoerceType blocks that impart it.
                         if isinstance(match_step.root_block, QueryRoot):
+                            # Question: What does it mean for the bound to be implied by the
+                            # query structure? I think we need an INSTANCEOF filter in this case.
```

Comment:
@obi1kenobi

Reply:
Oh hmm, I think in my haste to develop a fix for the query scheduling bug we observed in OrientDB that this code attempts to work around, I might have missed an edge case here. Good catch!

If you have an edge from types `A -> B`, and you have already enforced the type bound `A` before traversing the edge, then after traversing the edge (per the schema and query structure), you already know you are at a `B` and don't need an explicit `INSTANCEOF` check.

What your query points out, though, is an interesting edge case:
- The edge is defined as from types `A -> B`.
- `B` is a preferred location (due to its filter), so it runs first. The other side of the edge is an eligible location but not preferred.
- However, in the query, the class used on the "out" side of the edge is not `A` but a class `X` that is a subclass of `A`. This `X` is not implied by the schema and query structure, and requires an `INSTANCEOF` check. This check is not applied, producing incorrect results.

I would propose two actions here:
- The presence of a `Filter` that does not contain a `Variable` expression (i.e. filters that use tagged parameters instead of runtime parameters) should not make a location preferred, only eligible. This is because OrientDB (to my knowledge) is not able to take advantage of indexes in situations involving tagged parameters (i.e. `ContextField` expressions). This action does not remove the bug, but does remove an unintended consequence of my rushed implementation here, and I suspect will improve scheduling outcomes for a subset of queries.
- Improve the logic for eliminating `INSTANCEOF` checks on the basis of schema and query structure information, such that the edge case above is eliminated.

Reply:
My approach would be to rewrite this function. It's too hairy. The `previous_type_bound` and the concept of `implied by query structure` are just too vague and complicated. Too much is done at the same time, it's likely I will have an error in my fix.

Reply:
That's probably fine, have a go at it and let's see what it looks like.